app/src/main/java/com/vuongnq14798/vuongmp3/mediaplayer/MediaPlayerManager.kt

```diff
@@ -0,0 +1,137 @@
+    MediaPlayerInterface {
+
+    @StateType
+    var state: Int = StateType.PAUSED
+    @ShuffleType
+    var shuffleType = ShuffleType.OFF
+    @LoopType
+    var loopType = LoopType.NONE
+
+    private lateinit var currentTrack: Track
+
+    private val mediaPlayer = MediaPlayer()
+
+    private var tracks = arrayListOf<Track>()
+
+    override fun create(track: Track) {
+
+        mediaPlayer.reset()
+        mediaPlayer.setAudioAttributes(
+            AudioAttributes.Builder()
+                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
+                .build()
+        )
+
+        try {
+            mediaPlayer.apply {
+                setDataSource(mediaPlayerService, Uri.parse(track?.streamUrl))
+                setOnErrorListener(mediaPlayerService)
+                setOnCompletionListener(mediaPlayerService)
+                setOnPreparedListener(mediaPlayerService)
+                prepareAsync()
+            }
+
+        } catch (exception: Exception) {
+            mediaPlayerService.onFailure()
+        }
+    }
+
+    override fun start() {
+        state = StateType.PLAYING
+        mediaPlayer.start()
+    }
+
+    override fun pause() {
+        state = StateType.PAUSED
+        mediaPlayer.pause()
+    }
+
+    override fun stop() {
+        state = StateType.PLAYING
+        mediaPlayer.stop()
+    }
+
+    override fun release() = mediaPlayer.release()
+
+    override fun isPlaying(): Boolean = mediaPlayer.isPlaying
+
+    override fun seekTo(position: Int) = mediaPlayer.seekTo(position)
+
+    override fun getCurrentTrack(): Track = currentTrack
+
+    override fun setCurrentTrack(track: Track) {
+        currentTrack = track
+    }
+
+    override fun nextTrack() {
+        if (shuffleType == ShuffleType.OFF) changeTrack(getNextTrack())
+        else changeTrack(tracks[Random.nextInt(tracks.size)])
+    }
+
+    override fun previousTrack() {
+        if (shuffleType == ShuffleType.OFF) changeTrack(getPreviousTrack())
+        else changeTrack(tracks[Random.nextInt(tracks.size)])
+    }
+
+    override fun changeTrack(track: Track) {
+        if (tracks.indexOf(track) < 0) addTrack(track)
+        currentTrack = track;
+        create(track)
+    }
+
+    override fun getCurrentPosition(): Int = mediaPlayer.currentPosition
+
+    override fun isEndOfList(): Boolean =
+        currentTrack == null || tracks.indexOf(currentTrack) + 1 == tracks.size
+
+    override fun addTrack(track: Track) {
+        tracks.add(track)
+    }
+
+    override fun addTracks(listTracks: List<Track>) {
+        tracks.apply {
+            clear()
+            tracks.addAll(listTracks)
+        }
+    }
+
+    override fun removeTrack(track: Track) {
+        tracks.remove(track)
+    }
+
+    override fun clearTracks() {
+        tracks.clear()
+    }
+
+    private fun getNextTrack(): Track {
+        val position = tracks.indexOf(currentTrack)
+        return if (position == tracks.size - 1) tracks[0] else tracks[position + 1]
```

Comment:
Logic viết như này có được không nhỉ? 
mà mỗi lần nextTrack lại phải tìm kiếm lại từ đầu sao 
```suggestion
        return tracks.run {
            this[(position + 1 + size) % size]
        }
```